MFVNeutralino/plugins/Vertexer.cc

```diff
@@ -1,891 +1,1776 @@
+
+      // start collecting shared tracks of sv0 for each shared jet
+      std::multimap<int, size_t> sv0_m;
+      for (size_t k = 0; k < sv0_match_tracktojet_which_jetidx.size(); k++)
+        if (sv0_match_tracktojet_which_jetidx[k] == jet_index) { sv0_m.insert({ sv0_match_tracktojet_which_jetidx[k], k }); }
+
+      for (auto it = sv0_m.begin(); it != sv0_m.end(); )
+      {
+        auto p = sv0_m.equal_range(it->first);
+
+        while (p.first != p.second)
+        {
+
+          sv0_match_sharedjettotrack_which_temp_trkidx.push_back(sv0_match_tracktojet_which_trkidx[p.first++->second]);
+        }
+        it = p.second;
+
+      }
+
+      sv0_match_sharedjettotrack_which_trkidx.push_back(sv0_match_sharedjettotrack_which_temp_trkidx);
+
+      // start collecting shared tracks of sv1 for each shared jet
+      std::multimap<int, size_t> sv1_m;
+      for (size_t k = 0; k < sv1_match_tracktojet_which_jetidx.size(); k++)
+        if (sv1_match_tracktojet_which_jetidx[k] == jet_index) { sv1_m.insert({ sv1_match_tracktojet_which_jetidx[k], k }); }
+
+      for (auto it = sv1_m.begin(); it != sv1_m.end(); )
+      {
+        auto p = sv1_m.equal_range(it->first);
+
+        while (p.first != p.second)
+        {
+
+          sv1_match_sharedjettotrack_which_temp_trkidx.push_back(sv1_match_tracktojet_which_trkidx[p.first++->second]);
+        }
+        it = p.second;
+
+      }
+
+      sv1_match_sharedjettotrack_which_trkidx.push_back(sv1_match_sharedjettotrack_which_temp_trkidx);
+
+      if (sv0_match_sharedjettotrack_which_temp_trkidx.size() == 1) {
+        sv0_lonesharedtrack_which_trkidx.push_back(sv0_match_sharedjettotrack_which_temp_trkidx[0]);
+      }
+      if (sv1_match_sharedjettotrack_which_temp_trkidx.size() == 1) {
+        sv1_lonesharedtrack_which_trkidx.push_back(sv1_match_sharedjettotrack_which_temp_trkidx[0]);
+      }
+
+      sv0_match_sharedjettotrack_which_temp_trkidx = {};
+      sv1_match_sharedjettotrack_which_temp_trkidx = {};
+
+    }
+    sv_lonesharedtrack_which_trkidx.push_back(sv0_lonesharedtrack_which_trkidx);
+    sv_lonesharedtrack_which_trkidx.push_back(sv1_lonesharedtrack_which_trkidx);
+  }
+  bool shared_jet_specialcase = (sv0_lonesharedtrack_which_trkidx.size() != 0 || sv1_lonesharedtrack_which_trkidx.size() != 0);
+  return std::pair<bool, std::vector<std::vector<size_t>>>(
+      shared_jet_specialcase,
+      sv_lonesharedtrack_which_trkidx);
+}
+
+bool MFVVertexer::match_track_jet(const reco::Track& tk, const pat::Jet& matchjet, const pat::JetCollection& jets, const int& idx) {
+
+  if (verbose) {
+    std::cout << "jet track matching..." << std::endl;
+    std::cout << "  target track pt " << tk.pt() << " eta " << tk.eta() << " phi " << tk.phi() << std::endl;
+  }
+
+  // arbitrary threshold, but idea is to minimize [1+delta(pT)] * [1+delta(eta)] * [1+delta(phi)] in order to match the track to the jet

```

Comment:
I just added this comment, but a question for you is:
In practice, do you remember how often the valid match actually differs between the original track and the jet track? If I remember the logic, the idea behind 1.3 was to allow each measurement (pT, eta, phi) to be off by up to ~10% (where 1.1*1.1*1.1 = 1.331, which we rounded down to 1.3).

It was a bit of an arbitrary choice though. It may even be the case that valid matches agree to a much larger extent than that--it could be good to check, so that we know that our choice is still a good one.

Reply:
Again, not the function I wrote. Maybe @Ang-Li-95 knows?

Reply:
It is all copied from our initial studies of shared jets in VertexHistos (e.g. [here](https://github.com/DisplacedVertices/cmssw-usercode/blob/clean_vertexing/MFVNeutralino/plugins/VertexHistos.cc#L757-L775)), and I know the _history_ of why it is that way, but my question is more about whether or not it's the optimal choice for our usage in the shared jet mitigation. It should be a pretty quick check, and is important since our shared jet mitigation relies on picking the correct jets.

Reply:
We have checked the plot [here](https://pekotamn.web.cern.ch/pekotamn/NtupleV27m_test_match_vertexer_study/h_all_tracks_log.png) (from 30 tbs events) and it seems that the cut at 1.3 of the match threshold is still reasonable.